src/tools/pkcs15-tool.c

```diff
@@ -1515,11 +1515,47 @@ static int read_and_cache_file(const sc_path_t *path)
 		printf("out of memory!");
 		return -1;
 	}
-	r = sc_read_binary(card, 0, buf, size, 0);
-	if (r < 0) {
-		fprintf(stderr, "sc_read_binary() failed: %s\n", sc_strerror(r));
-		free(buf);
-		return -1;
+	if (tfile->ef_structure == SC_FILE_EF_LINEAR_VARIABLE_TLV) {
```

Comment:
some cards don't return metadata about the files so there is no way of knowing whether its a `SC_FILE_EF_LINEAR_VARIABLE_TLV`. This change would render pkcs15-tool unusable for many cards.